doc/compiler/optimizer/GlobalOptimizationsSummary.md

```diff
@@ -0,0 +1,138 @@
+OpenJDK Assembly Exception [2].
+
+[1] https://www.gnu.org/software/classpath/license.html
+[2] http://openjdk.java.net/legal/assembly-exception.html
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# Global Optimizations
+
+## Global [Value Propagation](https://github.com/eclipse/omr/blob/master/doc/compiler/optimizer/ValuePropagation.md)
+Global Value Propagation performs propagation of types, constants, ranges, nullness
+etc. for value numbers. It requires global value numbering to have been done.
+De-virtualization (followed by inlining in some cases), check removal, branch
+folding, replacement by constants (integer and strings) are consequences.
+
+## Partial Redundancy Elimination (PRE)
+Partial Redundancy Elimination performs global commoning and loop invariant code motion
+for normal expressions and exception checks. Based on Knoop, Ruthing and Steffen’s
+lazy code motion algorithm that involves two local analyses and five bit vector
+analyses. In our enhancement to the original algorithm to enable optimization of
+checks, we perform lazy code motion optimistically first (ignoring exception ordering
+constraints imposed by Java). We then perform a post-pass data flow analysis that attempts
+to move the expressions to their optimal placement points as computed by the traditional
+Least Common Multiple (LCM) algorithm. Note that it may not be possible to move an
+expression to the appropriate placement point as deemed by the optimistic analyses.
+
+## Escape Analysis (including allocation splitter)
+Escape Analysis uses value numbering to find all the uses of a particular allocation.
+If all uses of the allocated object are within the method, then the object is allocated
+on the stack. Synchronizations on non-escaping objects can also be eliminated. Some
+library methods are known to not cause the object to escape; these methods are special
+cased in the analysis.
+
+Stack allocations can be done in two forms: contiguous or non-contiguous.
+- Non-contiguous allocations represent the best case scenario from a performance viewpoint
+as the object and each of its constituent fields are represented by distinct temps.
+This enables each of the fields to be considered by other analyses (e.g. dead store etc.)
+completely independently of the others. Note that this can only be done if the allocation
+is not used as an object (e.g. passed out into a call which dereferences the object).
+
+- Contiguous allocations mean that the object is allocated on the stack but the fields
+are all part of the object on the stack. References to the fields still have to use
+the underlying object (except that it's on the stack instead of the heap).
+
+An early pass that splits (duplicates) one allocation into multiple allocations along
+different control flow paths (allocation splitting) is used to maximize the opportunities
+for escape analysis.
```

Comment:
I agree with @hzongaro There probably is an escape analysis talk and/or presentation from him that we could link to if it is in the open. 

Other points that we probably should mention : `There is a notion of "heapifying" a stack allocated object in cases even when an escape point is detected for the object but the escape point is a cold block. In such cases it is worth keeping the object on stack on the hot path and only if the cold path is taken, then the object is allocated on the heap and the contents of the object on the stack are copied into the newly allocated heap object. As part of heapification, all references from autos/parms to the stack allocated object are updated to point to the heapified object instead. This means that it is safe to execute any code where the object escapes from that point on.`

Reply:
EA overview -> https://youtu.be/JMpqwLVwmk8

Reply:
Other potential additions:

`Our escape analysis implementation also has a notion of de-memoization/re-memoization. This pertains to some special classes in the JCL: Integer, Long, Short, Byte etc. that memoize some instances of those classes in a static array that they maintain (only for some ranges of values). These classes would use these arrays to return an already allocated/memoized object rather than allocate a new object every time (if possible) when `valueOf` is called. These arrays therefore also become escape points for objects that are allocated and stored into them. De-memoization attempts to get around this by converting valueOf calls to new allocations before escape analysis is attempted on the assumption that allocating such objects on the stack would be even more preferable to memoizing them in the arrays. If stack allocation was not possible due to some reason, the de-memoized allocations are re-memoized, leaving the code as it was originally and an attempt is also made to inline the re-memoized valueOf since the inlining of these calls is suppressed in the inliner to give de-memoization a chance later.`

Reply:
Finally, on the Power platform (maybe AArch64 should use it too if it has a weak memory model, attn : @0xdaryl ) there is a "flow sensitive escape analysis" pass that gets run. 

`Flow sensitive escape analysis tries to find places in the code where an object "has not escaped yet" rather than the notion of "does the object escape anywhere" that is the hallmark of the flow insensitive escape analysis that runs on non-Power platforms. The reason for doing this flow sensitive analysis is because we want to optimize away the lwsync instructions that must be emitted after allocations to ensure that other threads see a properly initialized object. Flow sensitive escape analysis attempts to move such lysyncs together (i.e. combine) as much as possible. This includes trying to move an lwsync for an object to another lwsync (either for the same object (after a constructor if the object has final fields) or a different object), a monexit or volatile store (that do lwsyncs for different reasons) but all only on the basis that the object has not escaped yet.`